File: leetcode/C++/303.cpp
#include <vector>
using namespace std;

class NumArray
{
  public:
    vector<int> dp;
    NumArray(const vector<int> &nums)
    {
        dp.resize(nums.size());
        if (!nums.empty())
            dp[0] = nums[0];
        for (int i = 1; i < nums.size(); ++i)
        {
            dp[i] = nums[i] + dp[i - 1];
        }
    }

    int sumRange(int i, int j)
    {
        if (i == 0)
        {
            return dp[j];
        }
        return dp[j] - dp[i - 1];
    }
};

File: leetcode/Go/303.go
package main

type NumArray struct {
	dp []int
}


func Constructor(nums []int) NumArray {
	for i := 1; i < len(nums); i++ {
		nums[i] = nums[i-1]+nums[i]
	}
	return NumArray{nums}
}


func (this *NumArray) SumRange(i int, j int) int {
	if i == 0 {
		return this.dp[j]
	}
	return this.dp[j]-this.dp[i-1]
}
